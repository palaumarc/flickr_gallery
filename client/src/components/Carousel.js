import React, { Component } from 'react';
import './Carousel.css'

const DEFAULT_INDEX = 0;

class Carousel extends Component {

    getStartIndex = () => {
        const { children, startIndex } = this.props;
        if (startIndex === undefined || startIndex < 0 || startIndex >= React.Children.count(children)) {
            return DEFAULT_INDEX;
        }

        return startIndex;
    }

    selectNext = () => {
        const newIndex = this.state.selectedIndex + 1;
        this.setState({...this.state, selectedIndex: newIndex});
    }

    selectPrevious = () => {
        const newIndex = this.state.selectedIndex - 1;
        this.setState({...this.state, selectedIndex: newIndex});
    }

    state = {
        selectedIndex: this.getStartIndex()
    }

    render() {

        const reactChildrenAsArray = React.Children.toArray(this.props.children);

        const previousControl = (this.state.selectedIndex > 0) ?
            <div className="carousel-controls-previous" onClick={this.selectPrevious}>&lt;</div> : null;

        const nextControl = (this.state.selectedIndex < reactChildrenAsArray.length - 1) ?
            <div className="carousel-controls-next" onClick={this.selectNext}>&gt;</div> : null;

        return (
            <div className="carousel">
                <div className="carousel-content">
                    {reactChildrenAsArray[this.state.selectedIndex]}
                </div>
                <div className="carousel-controls">
                    {previousControl}
                    {nextControl}
                </div>
            </div>
        )

    }   

}

export default Carousel;