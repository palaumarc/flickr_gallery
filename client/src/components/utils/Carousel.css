.carousel {
    display: grid;
    grid-template-rows: 1fr 40px;
    height: 100%;
    justify-items: center;
}

.carousel-content {
    height: 100%;
    width: 100%;
}

.carousel-controls {
    width: 100px;
    align-self: center;
}

.carousel-controls-next, .carousel-controls-previous {
    color: white;
    font-size: 20px;
    font-weight: bold;
    cursor: pointer;
    z-index: 2;
}

.carousel-controls-next {
    float: right;
}

.carousel-controls-previous {
    float: left;
}

.carousel-controls-next:hover, .carousel-controls-previous:hover { 
    color: rgb(212, 212, 212);
}