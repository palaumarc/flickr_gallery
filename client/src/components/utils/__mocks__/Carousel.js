import React from "react";

const Carousel = props => <carousel-mock {...props}>{props.children}</carousel-mock>;

export default Carousel;
