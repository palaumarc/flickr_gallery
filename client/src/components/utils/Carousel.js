import React, { Component } from 'react';
import PropTypes from 'prop-types';

import './Carousel.css'

class Carousel extends Component {

    static propTypes = {
        startIndex: PropTypes.number,
        children: PropTypes.node.isRequired,
    }

    static defaultProps = {
        startIndex: 0
    }

    state = {
        selectedIndex: this.props.startIndex
    }

    selectNext = () => {
        this.setState(prevState => ({...prevState, selectedIndex: prevState.selectedIndex + 1}));
    }

    selectPrevious = () => {
        this.setState(prevState => ({...prevState, selectedIndex: prevState.selectedIndex - 1}));
    }

    render() {

        const reactChildrenAsArray = React.Children.toArray(this.props.children);

        const previousControl = (this.state.selectedIndex > 0) ?
            <div className="carousel-controls-previous" onClick={this.selectPrevious}>&lt;</div> : null;

        const nextControl = (this.state.selectedIndex < reactChildrenAsArray.length - 1) ?
            <div className="carousel-controls-next" onClick={this.selectNext}>&gt;</div> : null;

        return (
            <div className="carousel">
                <div className="carousel-content">
                    {reactChildrenAsArray[this.state.selectedIndex]}
                </div>
                <div className="carousel-controls">
                    {previousControl}
                    {nextControl}
                </div>
            </div>
        )

    }   

}

export default Carousel;