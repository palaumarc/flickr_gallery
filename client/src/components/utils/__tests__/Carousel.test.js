import React from "react";
import renderer from "react-test-renderer";
import Carousel from "../Carousel";

describe('Component is rendered four children and with prop startIndex 2.', () => {

    const componentToTest = (
        <Carousel startIndex={2}>
            <div>Carousel child 0</div>
            <div>Carousel child 1</div>
            <div>Carousel child 2</div>
            <div>Carousel child 3</div>
        </Carousel>
    );

    test('Snapshot', () => {
        const tree = renderer.create(componentToTest);
        expect(tree.toJSON()).toMatchSnapshot();
    });

    test('Next button is clicked. Snapshot', () => {
        const tree = renderer.create(componentToTest);
        const nextButton = tree.root.find(element => element.children == '>');
        nextButton.props.onClick();
        expect(tree.toJSON()).toMatchSnapshot();
    });

    test('Previous button is clicked. Snapshot', () => {
        const tree = renderer.create(componentToTest);
        const previousButton = tree.root.find(element => element.children == '<');
        previousButton.props.onClick();
        expect(tree.toJSON()).toMatchSnapshot();
    });

}) 

describe('Component is rendered with four children and without prop startIndex', () => {

    const componentToTest = (
        <Carousel>
            <div>Carousel child 0</div>
            <div>Carousel child 1</div>
            <div>Carousel child 2</div>
            <div>Carousel child 3</div>
        </Carousel>
    );

    test('Snapshot', () => {
        const tree = renderer.create(componentToTest);
        expect(tree.toJSON()).toMatchSnapshot();
    });

})