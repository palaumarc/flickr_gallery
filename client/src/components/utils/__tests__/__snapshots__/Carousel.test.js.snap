// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component is rendered four children and with prop startIndex 2. Next button is clicked. Snapshot 1`] = `
<div
  className="carousel"
>
  <div
    className="carousel-content"
  >
    <div>
      Carousel child 3
    </div>
  </div>
  <div
    className="carousel-controls"
  >
    <div
      className="carousel-controls-previous"
      onClick={[Function]}
    >
      &lt;
    </div>
  </div>
</div>
`;

exports[`Component is rendered four children and with prop startIndex 2. Previous button is clicked. Snapshot 1`] = `
<div
  className="carousel"
>
  <div
    className="carousel-content"
  >
    <div>
      Carousel child 1
    </div>
  </div>
  <div
    className="carousel-controls"
  >
    <div
      className="carousel-controls-previous"
      onClick={[Function]}
    >
      &lt;
    </div>
    <div
      className="carousel-controls-next"
      onClick={[Function]}
    >
      &gt;
    </div>
  </div>
</div>
`;

exports[`Component is rendered four children and with prop startIndex 2. Snapshot 1`] = `
<div
  className="carousel"
>
  <div
    className="carousel-content"
  >
    <div>
      Carousel child 2
    </div>
  </div>
  <div
    className="carousel-controls"
  >
    <div
      className="carousel-controls-previous"
      onClick={[Function]}
    >
      &lt;
    </div>
    <div
      className="carousel-controls-next"
      onClick={[Function]}
    >
      &gt;
    </div>
  </div>
</div>
`;

exports[`Component is rendered with four children and without prop startIndex Snapshot 1`] = `
<div
  className="carousel"
>
  <div
    className="carousel-content"
  >
    <div>
      Carousel child 0
    </div>
  </div>
  <div
    className="carousel-controls"
  >
    <div
      className="carousel-controls-next"
      onClick={[Function]}
    >
      &gt;
    </div>
  </div>
</div>
`;
